tests/ranged_balance.cpp

```diff
@@ -0,0 +1,123 @@
+                                           const tripoint &target_arg, dispersion_sources dispersion,
+                                           Creature *origin, const vehicle *in_veh );
+
+static void equip_shooter( npc &shooter, itype_id gun_id )
+{
+    tripoint shooter_pos( 60, 60, 0 );
+    shooter.setpos( shooter_pos );
+    shooter.worn.clear();
+    shooter.inv.clear();
+    shooter.remove_weapon();
+    // So we don't drop anything.
+    shooter.wear_item( item( "backpack" ) );
+
+    // Give shooter a loaded gun of the requested type.
+    item &gun = shooter.i_add( item( gun_id ) );
+    const itype_id ammo_id = gun.ammo_default();
+    if( gun.magazine_integral() ) {
+        item &ammo = shooter.i_add( item( ammo_id, calendar::turn, gun.ammo_capacity() ) );
+	REQUIRE( gun.is_reloadable_with( ammo_id ) );
+	REQUIRE( shooter.can_reload( gun, ammo_id ) );
+	gun.reload( shooter, item_location( shooter, &ammo ), gun.ammo_capacity() );
+    } else {
+        const itype_id magazine_id = gun.magazine_default();
+        item &magazine = shooter.i_add( item( magazine_id ) );
+        item &ammo = shooter.i_add( item( ammo_id, calendar::turn, magazine.ammo_capacity() ) );
+        REQUIRE( magazine.is_reloadable_with( ammo_id ) );
+        REQUIRE( shooter.can_reload( magazine, ammo_id ) );
+        magazine.reload( shooter, item_location( shooter, &ammo ), magazine.ammo_capacity() );
+        gun.reload( shooter, item_location( shooter, &magazine ), magazine.ammo_capacity() );
+    }
+    shooter.wield( gun );
+}
+
+static statistics firing_test( npc &shooter, std::string gun_type, float aim_ratio, int range )
+{
+    equip_shooter( shooter, itype_id( gun_type ) );
+
+    // Spawn a target.
+    // Maybe change this to a random direction, or even move the target around during the test.
+    tripoint target_pos( shooter.pos() + tripoint( range, 0, 0 ) );
+    monster &test_target = spawn_test_monster( "debug_mon", target_pos );
+
+    item &gun = shooter.weapon;
+
+    dispersion_sources dispersion =
+      shooter.get_weapon_dispersion( gun,
+                                     rl_dist( shooter.pos(), test_target.pos() ) );
+    dispersion.add_range( std::max( static_cast<double>( gun.sight_dispersion() ),
+                                    shooter.recoil_total() * ( 1.0 - aim_ratio ) ) );
+
+    statistics firing_stats;
+    do {
+        // Top off target hp to make sure it doesn't die.
+        test_target.set_hp( test_target.get_hp_max() );
+        // Calling projectile_ttack() instead of fire_gun() because it returns wether it was a hit or not.
+        // Conviniently, it also avoids consuming the ammo.
+        dealt_projectile_attack shot =
+          projectile_attack( make_gun_projectile( gun ), shooter.pos(), test_target.pos(),
+                             dispersion, &shooter, nullptr );
+        firing_stats.add( shot.hit_critter != nullptr );
```

Comment:
If it's practical to make assertions about hit quality, I'll also make assertions about it, but grazes aren't meaningless to make assertions about.